diff --git a/quickjs.c b/quickjs.c
index c0aaf23..d0aac4a 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -42487,6 +42487,8 @@ static int BC_add_object_ref1(BCReaderState *s, JSObject *p)
                             &s->objects_size, s->objects_count + 1))
             return -1;
         s->objects[s->objects_count++] = p;
+        if (p)
+            JS_REF_COUNT(p)++;
     }
     return 0;
 }
@@ -43706,6 +43708,7 @@ static JSValue JS_ReadTypedArray(BCReaderState *s)
         goto fail;
     if (s->allow_reference) {
         s->objects[idx] = JS_VALUE_GET_OBJ(obj);
+        JS_REF_COUNT(JS_VALUE_GET_OBJ(obj))++;
     }
     JS_FreeValue(ctx, array_buffer);
     return obj;
@@ -44149,6 +44152,11 @@ static void bc_reader_free(BCReaderState *s)
         }
         js_free(s->ctx, s->idx_to_atom);
     }
+    for (i = 0; i < s->objects_count; i++) {
+        if (s->objects[i])
+            JS_FreeValue(s->ctx,
+                         JS_MKPTR(JS_TAG_OBJECT, s->objects[i]));
+    }
     js_free(s->ctx, s->objects);
     js_free(s->ctx, s->var_refs);
     for (i = 0; i < s->weak_targets_count; i++)
