diff --git a/quickjs.c b/quickjs.c
index 3cbbc27..5745ba9 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -30365,7 +30365,7 @@ static __exception int js_parse_statement(JSParseState *s)
     return js_parse_statement_or_decl(s, 0);
 }
 
-static __exception int js_parse_block(JSParseState *s)
+static __exception int js_parse_block(JSParseState *s, uint32_t *end_pos)
 {
     JSFunctionDef *fd = s->cur_func;
 
@@ -30416,6 +30416,8 @@ static __exception int js_parse_block(JSParseState *s)
         }
         pop_scope(s);
     }
+    if (end_pos)
+        *end_pos = s->last_ptr - s->buf_start;
     if (next_token(s))
         return -1;
     return 0;
@@ -31034,7 +31036,7 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
 
     switch(tok = s->token.val) {
     case '{':
-        if (js_parse_block(s))
+        if (js_parse_block(s, NULL))
             goto fail;
         break;
     case TOK_RETURN:
@@ -31618,7 +31620,7 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
                              JS_ATOM_NULL, -1, -1, 1);
             block_env.label_finally = label_finally;
 
-            if (js_parse_block(s))
+            if (js_parse_block(s, &statement_start))
                 goto fail;
 
             pop_break_entry(s->cur_func);
@@ -31640,6 +31642,7 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
 
                 push_scope(s);  /* catch variable */
                 emit_label(s, label_catch);
+                coverage_pos = emit_coverage(s, statement_start);
 
                 if (s->token.val == '{') {
                     /* support optional-catch-binding feature */
@@ -31679,8 +31682,10 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
                                  -1, -1, 1);
                 block_env.label_finally = label_finally;
 
-                if (js_parse_block(s))
+                if (js_parse_block(s, NULL))
                     goto fail;
+                close_coverage(s, coverage_pos,
+                               s->last_ptr - s->buf_start);
 
                 pop_break_entry(s->cur_func);
                 pop_scope(s);  /* catch block */
@@ -31740,7 +31745,7 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
                     set_eval_ret_undefined(s);
                 }
 
-                if (js_parse_block(s))
+                if (js_parse_block(s, NULL))
                     goto fail;
                 close_coverage(s, coverage_pos,
                                s->last_ptr - s->buf_start);
