diff --git a/quickjs.c b/quickjs.c
index f1e6127..143ee0d 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -8408,7 +8408,7 @@ static void js_coverage_register_function(JSFunctionBytecode *b)
     const char *filename = "";
     bool free_function_name = false;
     bool free_filename = false;
-    int line_num, column_num, pc, value, ret;
+    int i, line_num, column_num, pc, value, ret;
     unsigned int op;
 
     if (b->func_name != JS_ATOM_NULL) {
@@ -8427,6 +8427,11 @@ static void js_coverage_register_function(JSFunctionBytecode *b)
     if (free_filename)
         JS_FreeCString(b->realm, filename);
 
+    for (i = 0; i < b->cpool_count; i++) {
+        if (JS_VALUE_GET_TAG(b->cpool[i]) == JS_TAG_FUNCTION_BYTECODE)
+            js_coverage_register_function(JS_VALUE_GET_PTR(b->cpool[i]));
+    }
+
     line_num = b->line_num;
     column_num = b->col_num;
     v82jsc_coverage_location(b, line_num, column_num);
