--- a/quickjs.c
+++ b/quickjs.c
@@ -28016,5 +28016,10 @@ static int js_parse_destructuring_element(JSParseState *s, int tok,
                     if (js_define_var(s, var_name, tok))
                         goto var_error;
+                    if (export_flag) {
+                        if (!add_export_entry(s, s->cur_func->module, var_name, var_name,
+                                              JS_EXPORT_TYPE_LOCAL))
+                            goto var_error;
+                    }
                     opcode = OP_scope_get_var;
                     scope = s->cur_func->scope_level;
                 } else {
