diff --git a/quickjs.c b/quickjs.c
index 7b2f99e..979b96e 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -37881,7 +37881,7 @@ static int add_module_variables(JSContext *ctx, JSFunctionDef *fd)
             idx = find_closure_var(ctx, fd, me->local_name);
             if (idx < 0) {
                 // XXX: add_module_variables() should take JSParseState *s and use js_parse_error_atom
-                JS_ThrowSyntaxErrorAtom(ctx, "exported variable '%s' does not exist",
+                JS_ThrowSyntaxErrorAtom(ctx, "Export '%s' is not defined in module",
                                         me->local_name);
                 return -1;
             }
