--- a/quickjs.c
+++ b/quickjs.c
@@ -29561,6 +29561,9 @@
         emit_op(s, OP_drop);
         emit_op(s, OP_drop);
     }
+    emit_op(s, OP_dup);
+    emit_op(s, OP_is_undefined);
+    label_done = emit_goto(s, OP_if_true, -1);
     emit_op(s, OP_get_field2);
     emit_atom(s, JS_ATOM_return);
     /* completion? iter_obj return_func */
@@ -29571,7 +29574,7 @@
     emit_u16(s, 0);
     emit_op(s, OP_await);
     emit_op(s, OP_check_object);
-    label_done = emit_goto(s, OP_goto, -1);
+    emit_goto(s, OP_goto, label_done);
     emit_label(s, label_no_return);
     emit_op(s, OP_drop);
     emit_label(s, label_done);
