diff --git a/quickjs.c b/quickjs.c
--- a/quickjs.c
+++ b/quickjs.c
@@ -41326,8 +41326,9 @@ static int JS_WriteObjectAtoms(BCWriterState *s)
             bc_put_u32(s, atom);
         } else {
             JSAtomStruct *p = rt->atom_array[atom];
-            uint8_t type = p->atom_type;
-            assert(type != JS_ATOM_TYPE_PRIVATE);
+            uint8_t type = p->atom_type == JS_ATOM_TYPE_SYMBOL &&
+                           p->hash == JS_ATOM_HASH_PRIVATE
+                               ? JS_ATOM_TYPE_PRIVATE : p->atom_type;
             bc_put_u8(s, type);
             JS_WriteString(s, p);
         }
@@ -43414,7 +43415,7 @@ static int JS_ReadObjectAtoms(BCReaderState *s)
                 return -1;
             }
         } else {
-            if (type < JS_ATOM_TYPE_STRING || type >= JS_ATOM_TYPE_PRIVATE) {
+            if (type < JS_ATOM_TYPE_STRING || type > JS_ATOM_TYPE_PRIVATE) {
                 JS_ThrowInternalError(s->ctx, "invalid symbol type %d", type);
                 return -1;
             }
