diff --git a/quickjs.c b/quickjs.c
index 5745ba9..5125689 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -31126,7 +31126,7 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
     case TOK_IF:
         {
             int label1, label2, mask, coverage_pos;
-            uint32_t branch_start = s->token.ptr - s->buf_start;
+            uint32_t branch_start;
             if (next_token(s))
                 goto fail;
             /* create a new scope for `let f;if(1) function f(){}` */
@@ -31134,6 +31134,7 @@ static __exception int js_parse_statement_or_decl(JSParseState *s,
             set_eval_ret_undefined(s);
             if (js_parse_expr_paren(s))
                 goto fail;
+            branch_start = s->token.ptr - s->buf_start;
             label1 = emit_goto(s, OP_if_false, -1);
             if (s->cur_func->is_strict_mode)
                 mask = 0;
