diff --git a/quickjs.c b/quickjs.c
index 8b704c2..ee36220 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -9129,6 +9129,19 @@ static int JS_ThrowTypeErrorReadOnly(JSContext *ctx, int flags, JSAtom atom)
     }
 }
 
+static int JS_ThrowTypeErrorNotExtensible(JSContext *ctx, int flags,
+                                          JSAtom atom)
+{
+    if ((flags & JS_PROP_THROW) ||
+        ((flags & JS_PROP_THROW_STRICT) && is_strict_mode(ctx))) {
+        JS_ThrowTypeErrorAtom(ctx,
+                              "Cannot add property %s, object is not extensible",
+                              atom);
+        return -1;
+    }
+    return false;
+}
+
 JSValue JS_ThrowOutOfMemory(JSContext *ctx)
 {
     JSRuntime *rt = ctx->rt;
@@ -11357,7 +11370,7 @@ retry:
 
     if (p == JS_VALUE_GET_OBJ(obj)) {
         if (unlikely(!p->extensible)) {
-            ret = JS_ThrowTypeErrorOrFalse(ctx, flags, "object is not extensible");
+            ret = JS_ThrowTypeErrorNotExtensible(ctx, flags, prop);
             goto done;
         }
         if (p->is_exotic) {
@@ -11401,7 +11414,7 @@ retry:
                                 JS_PROP_HAS_VALUE);
     } else {
         if (unlikely(!p->extensible)) {
-            ret = JS_ThrowTypeErrorOrFalse(ctx, flags, "object is not extensible");
+            ret = JS_ThrowTypeErrorNotExtensible(ctx, flags, prop);
             goto done;
         }
     generic_create_prop:
