diff --git a/quickjs.c b/quickjs.c
index 531a774..c777e54 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -1541,7 +1541,8 @@ static JSValue js_new_promise_capability(JSContext *ctx,
 static __exception int perform_promise_then(JSContext *ctx,
                                             JSValueConst promise,
                                             JSValueConst *resolve_reject,
-                                            JSValueConst *cap_resolving_funcs);
+                                            JSValueConst *cap_resolving_funcs,
+                                            JSValueConst hook_promise);
 static JSValue js_promise_resolve(JSContext *ctx, JSValueConst this_val,
                                   int argc, JSValueConst *argv, int magic);
 static JSValue js_promise_then(JSContext *ctx, JSValueConst this_val,
@@ -22828,6 +22829,7 @@ static bool js_async_function_resume(JSContext *ctx, JSAsyncFunctionData *s)
             goto resolved;
         } else {
             JSValue promise, resolving_funcs[2], resolving_funcs1[2];
+            JSValue hook_promise;
             int i, res;
 
             /* await */
@@ -22844,16 +22846,30 @@ static bool js_async_function_resume(JSContext *ctx, JSAsyncFunctionData *s)
             set_stack_frame_gc_owner(ctx->rt, &s->func_state.frame,
                                      resolving_funcs[0]);
 
-            /* Note: no need to create 'thrownawayCapability' as in
-               the spec */
             for(i = 0; i < 2; i++)
                 resolving_funcs1[i] = JS_UNDEFINED;
+            hook_promise = JS_UNDEFINED;
+            if (ctx->rt->promise_hook) {
+                JSValueLink link = { ctx->rt->parent_promise, promise };
+                ctx->rt->parent_promise = &link;
+                hook_promise = JS_NewPromiseCapability(ctx, resolving_funcs1);
+                ctx->rt->parent_promise = link.next;
+                if (JS_IsException(hook_promise)) {
+                    JS_FreeValue(ctx, promise);
+                    for(i = 0; i < 2; i++)
+                        JS_FreeValue(ctx, resolving_funcs[i]);
+                    goto fail;
+                }
+            }
             res = perform_promise_then(ctx, promise,
                                        vc(resolving_funcs),
-                                       vc(resolving_funcs1));
+                                       vc(resolving_funcs1), hook_promise);
+            JS_FreeValue(ctx, hook_promise);
             JS_FreeValue(ctx, promise);
             for(i = 0; i < 2; i++)
                 JS_FreeValue(ctx, resolving_funcs[i]);
+            for(i = 0; i < 2; i++)
+                JS_FreeValue(ctx, resolving_funcs1[i]);
             if (res)
                 goto fail;
         }
@@ -23116,6 +23132,7 @@ static int js_async_generator_await(JSContext *ctx,
                                     JSValue value)
 {
     JSValue promise, resolving_funcs[2], resolving_funcs1[2];
+    JSValue hook_promise;
     int i, res;
 
     promise = js_promise_resolve(ctx, ctx->promise_ctor,
@@ -23129,16 +23146,30 @@ static int js_async_generator_await(JSContext *ctx,
         goto fail;
     }
 
-    /* Note: no need to create 'thrownawayCapability' as in
-       the spec */
     for(i = 0; i < 2; i++)
         resolving_funcs1[i] = JS_UNDEFINED;
+    hook_promise = JS_UNDEFINED;
+    if (ctx->rt->promise_hook) {
+        JSValueLink link = { ctx->rt->parent_promise, promise };
+        ctx->rt->parent_promise = &link;
+        hook_promise = JS_NewPromiseCapability(ctx, resolving_funcs1);
+        ctx->rt->parent_promise = link.next;
+        if (JS_IsException(hook_promise)) {
+            JS_FreeValue(ctx, promise);
+            for(i = 0; i < 2; i++)
+                JS_FreeValue(ctx, resolving_funcs[i]);
+            goto fail;
+        }
+    }
     res = perform_promise_then(ctx, promise,
                                vc(resolving_funcs),
-                               vc(resolving_funcs1));
+                               vc(resolving_funcs1), hook_promise);
+    JS_FreeValue(ctx, hook_promise);
     JS_FreeValue(ctx, promise);
     for(i = 0; i < 2; i++)
         JS_FreeValue(ctx, resolving_funcs[i]);
+    for(i = 0; i < 2; i++)
+        JS_FreeValue(ctx, resolving_funcs1[i]);
     if (res)
         goto fail;
     return 0;
@@ -23225,7 +23256,7 @@ static int js_async_generator_completed_return(JSContext *ctx,
     resolving_funcs[1] = JS_UNDEFINED;
     res = perform_promise_then(ctx, promise,
                                vc(resolving_funcs1),
-                               vc(resolving_funcs));
+                               vc(resolving_funcs), JS_UNDEFINED);
     JS_FreeValue(ctx, resolving_funcs1[0]);
     JS_FreeValue(ctx, resolving_funcs1[1]);
     JS_FreeValue(ctx, promise);
@@ -58786,6 +58817,7 @@ typedef struct JSPromiseReactionData {
     JSValue resolving_funcs[2];
     JSValue handler;
     JSValue job_context;
+    JSValue hook_promise;
 } JSPromiseReactionData;
 
 JSPromiseStateEnum JS_PromiseState(JSContext *ctx, JSValueConst promise)
@@ -58865,6 +58897,7 @@ static void promise_reaction_data_free(JSRuntime *rt,
     JS_FreeValueRT(rt, rd->resolving_funcs[1]);
     JS_FreeValueRT(rt, rd->handler);
     JS_FreeValueRT(rt, rd->job_context);
+    JS_FreeValueRT(rt, rd->hook_promise);
     js_free_rt(rt, rd);
 }
 
@@ -58884,6 +58917,8 @@ static JSValue promise_reaction_job(JSContext *ctx, int argc,
     JSValueConst handler, func;
     JSValue res, res2;
     JSValueConst arg;
+    JSPromiseHook *promise_hook;
+    void *promise_hook_opaque;
     bool is_reject;
 
     assert(argc == 6);
@@ -58893,9 +58928,13 @@ static JSValue promise_reaction_job(JSContext *ctx, int argc,
 
     promise_trace(ctx, "promise_reaction_job: is_reject=%d\n", is_reject);
 
-    if (ctx->rt->promise_hook) {
-        ctx->rt->promise_hook(ctx, JS_PROMISE_HOOK_BEFORE, argv[5],
-                              JS_UNDEFINED, ctx->rt->promise_hook_opaque);
+    promise_hook = ctx->rt->promise_hook;
+    promise_hook_opaque = ctx->rt->promise_hook_opaque;
+    if (promise_hook && JS_IsPromise(argv[5])) {
+        promise_hook(ctx, JS_PROMISE_HOOK_BEFORE, argv[5],
+                     JS_UNDEFINED, promise_hook_opaque);
+    } else {
+        promise_hook = NULL;
     }
     if (JS_IsUndefined(handler)) {
         if (is_reject) {
@@ -58906,14 +58945,15 @@ static JSValue promise_reaction_job(JSContext *ctx, int argc,
     } else {
         res = JS_Call(ctx, handler, JS_UNDEFINED, 1, &arg);
     }
-    if (ctx->rt->promise_hook) {
-        ctx->rt->promise_hook(ctx, JS_PROMISE_HOOK_AFTER, argv[5],
-                              JS_UNDEFINED, ctx->rt->promise_hook_opaque);
-    }
     is_reject = JS_IsException(res);
     if (is_reject) {
-        if (unlikely(JS_IsUncatchableError(ctx->rt->current_exception)))
+        if (unlikely(JS_IsUncatchableError(ctx->rt->current_exception))) {
+            if (promise_hook) {
+                promise_hook(ctx, JS_PROMISE_HOOK_AFTER, argv[5],
+                             JS_UNDEFINED, promise_hook_opaque);
+            }
             return JS_EXCEPTION;
+        }
         res = JS_GetException(ctx);
     }
     func = argv[is_reject];
@@ -58927,6 +58967,11 @@ static JSValue promise_reaction_job(JSContext *ctx, int argc,
     }
     JS_FreeValue(ctx, res);
 
+    if (promise_hook) {
+        promise_hook(ctx, JS_PROMISE_HOOK_AFTER, argv[5],
+                     JS_UNDEFINED, promise_hook_opaque);
+    }
+
     return res2;
 }
 
@@ -58959,7 +59004,7 @@ static void fulfill_or_reject_promise(JSContext *ctx, JSValueConst promise,
 
     promise_trace(ctx, "fulfill_or_reject_promise: is_reject=%d\n", is_reject);
 
-    if (s->promise_state == JS_PROMISE_FULFILLED) {
+    {
         JSRuntime *rt = ctx->rt;
         if (rt->promise_hook) {
             rt->promise_hook(ctx, JS_PROMISE_HOOK_RESOLVE, promise,
@@ -58976,7 +59021,7 @@ static void fulfill_or_reject_promise(JSContext *ctx, JSValueConst promise,
         args[2] = rd->handler;
         args[3] = js_bool(is_reject);
         args[4] = value;
-        args[5] = promise;
+        args[5] = rd->hook_promise;
         js_enqueue_job(ctx, promise_reaction_job, 6, args,
                        &rd->job_context);
         list_del(&rd->link);
@@ -59187,6 +59232,7 @@ static void js_promise_mark(JSRuntime *rt, JSValueConst val,
             JS_MarkValue(rt, rd->resolving_funcs[1], mark_func);
             JS_MarkValue(rt, rd->handler, mark_func);
             JS_MarkValue(rt, rd->job_context, mark_func);
+            JS_MarkValue(rt, rd->hook_promise, mark_func);
         }
     }
     JS_MarkValue(rt, s->promise_result, mark_func);
@@ -59352,7 +59398,8 @@ JSValue JS_PromiseThen(JSContext *ctx, JSValueConst promise,
     if (JS_IsException(result_promise))
         return result_promise;
 
-    ret = perform_promise_then(ctx, promise, handlers, vc(resolving_funcs));
+    ret = perform_promise_then(ctx, promise, handlers, vc(resolving_funcs),
+                               result_promise);
     for (i = 0; i < 2; i++)
         JS_FreeValue(ctx, resolving_funcs[i]);
     if (ret) {
@@ -59796,7 +59843,8 @@ static JSValue js_promise_race(JSContext *ctx, JSValueConst this_val,
 static __exception int perform_promise_then(JSContext *ctx,
                                             JSValueConst promise,
                                             JSValueConst *resolve_reject,
-                                            JSValueConst *cap_resolving_funcs)
+                                            JSValueConst *cap_resolving_funcs,
+                                            JSValueConst hook_promise)
 {
     JSPromiseData *s = JS_GetOpaque(promise, JS_CLASS_PROMISE);
     JSPromiseReactionData *rd_array[2], *rd;
@@ -59824,6 +59872,7 @@ static __exception int perform_promise_then(JSContext *ctx,
         } else {
             rd->job_context = JS_UNDEFINED;
         }
+        rd->hook_promise = js_dup(hook_promise);
         rd_array[i] = rd;
     }
 
@@ -59840,7 +59889,7 @@ static __exception int perform_promise_then(JSContext *ctx,
         args[2] = rd->handler;
         args[3] = js_bool(i);
         args[4] = s->promise_result;
-        args[5] = promise;
+        args[5] = rd->hook_promise;
         js_enqueue_job(ctx, promise_reaction_job, 6, args,
                        &rd->job_context);
         for(i = 0; i < 2; i++)
@@ -59880,7 +59929,8 @@ static JSValue js_promise_then(JSContext *ctx, JSValueConst this_val,
     JS_FreeValue(ctx, ctor);
     if (JS_IsException(result_promise))
         return result_promise;
-    ret = perform_promise_then(ctx, this_val, argv, vc(resolving_funcs));
+    ret = perform_promise_then(ctx, this_val, argv, vc(resolving_funcs),
+                               result_promise);
     for(i = 0; i < 2; i++)
         JS_FreeValue(ctx, resolving_funcs[i]);
     if (ret) {
@@ -60175,7 +60225,7 @@ static JSValue js_async_from_sync_iterator_next(JSContext *ctx, JSValueConst thi
 
         res = perform_promise_then(ctx, value_wrapper_promise,
                                    vc(resolve_reject),
-                                   vc(resolving_funcs));
+                                   vc(resolving_funcs), promise);
         JS_FreeValue(ctx, resolve_reject[0]);
         JS_FreeValue(ctx, value_wrapper_promise);
         JS_FreeValue(ctx, resolving_funcs[0]);
diff --git a/quickjs.h b/quickjs.h
index 5c6b532..3a40553 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -1156,7 +1156,7 @@ typedef enum JSPromiseHookType {
     JS_PROMISE_HOOK_INIT,     // emitted when a new promise is created
     JS_PROMISE_HOOK_BEFORE,   // runs right before promise.then is invoked
     JS_PROMISE_HOOK_AFTER,    // runs right after promise.then is invoked
-    JS_PROMISE_HOOK_RESOLVE,  // not emitted for rejected promises
+    JS_PROMISE_HOOK_RESOLVE,  // emitted when a promise is settled
 } JSPromiseHookType;
 
 // parent_promise is only passed in when type == JS_PROMISE_HOOK_INIT and
