diff --git a/core/iwasm/interpreter/wasm_interp_classic.c b/core/iwasm/interpreter/wasm_interp_classic.c
--- a/core/iwasm/interpreter/wasm_interp_classic.c
+++ b/core/iwasm/interpreter/wasm_interp_classic.c
@@ -7577,9 +7577,7 @@ wasm_interp_call_wasm(WASMExecEnv *exec_env,
     }
     else {
 #if WASM_ENABLE_DUMP_CALL_STACK != 0
-        if (wasm_interp_create_call_stack(exec_env)) {
-            wasm_interp_dump_call_stack(exec_env, true, NULL, 0);
-        }
+        wasm_interp_create_call_stack(exec_env);
 #endif
     }
 
diff --git a/core/iwasm/interpreter/wasm_interp_fast.c b/core/iwasm/interpreter/wasm_interp_fast.c
--- a/core/iwasm/interpreter/wasm_interp_fast.c
+++ b/core/iwasm/interpreter/wasm_interp_fast.c
@@ -8009,9 +8009,7 @@ wasm_interp_call_wasm(WASMExecEnv *exec_env,
     }
     else {
 #if WASM_ENABLE_DUMP_CALL_STACK != 0
-        if (wasm_interp_create_call_stack(exec_env)) {
-            wasm_interp_dump_call_stack(exec_env, true, NULL, 0);
-        }
+        wasm_interp_create_call_stack(exec_env);
 #endif
     }
 
diff --git a/core/iwasm/interpreter/wasm_runtime.c b/core/iwasm/interpreter/wasm_runtime.c
--- a/core/iwasm/interpreter/wasm_runtime.c
+++ b/core/iwasm/interpreter/wasm_runtime.c
@@ -3667,8 +3667,6 @@ wasm_interp_call_func_native(WASMModuleInstance *module_inst,
      * then the stack-frame is already freed inside wasm_interp_call_wasm */
     if (!ret) {
 #if WASM_ENABLE_DUMP_CALL_STACK != 0
-        if (wasm_interp_create_call_stack(exec_env)) {
-            wasm_interp_dump_call_stack(exec_env, true, NULL, 0);
-        }
+        wasm_interp_create_call_stack(exec_env);
 #endif
     }
